### ISSUE TITLE
False positives related to data readers

### ISSUE THREAD

**Issue by marcobiscaro2112** [clj-kondo/clj-kondo#2134]
**version**

```
$ clj-kondo --version
clj-kondo v2023.05.26
```

**macro usage**

No.

**platform**

```
$ sw_vers
ProductName:            macOS
ProductVersion:         13.4.1
BuildVersion:           22F82

$ file $(which clj-kondo)
/usr/local/bin/clj-kondo: Mach-O 64-bit executable x86_64
```

**editor**

N/A.

**problem**

When using custom data readers with a private reader function, clj-kondo reports an error when referring to the private var in `data_readers.clj` (which is allowed by Clojure), and also reports an warning that the private var is unused (which is not the case).

Using a private function as reader is useful when a library doesn't want to publicly expose the reader function as part of the API. 

Also, I couldn't find a way to workaround the linter error. Clojure disallows using the var syntax in `data_readers.clj` (e.g. `#'ns/var`) and apparently there is no way to change the linter configuration only for the `data_readers.clj` file (it is not a namespace). Adding a `:clj-kondo/ignore` comment directly to the data readers map is undesirable, because this resource file is part of the release artifact, and linter comments leaking to consumers of the library not ideal.

**repro**

```
$ cat project.clj 
(defproject private-readers "0.1.0-SNAPSHOT"
  :dependencies [[org.clojure/clojure "1.11.1"]])

$ cat src/private_readers/core.clj 
(ns private-readers.core)
(defn- my-tag-reader [x] x)

$ cat resources/data_readers.clj 
{my/tag private-readers.core/my-tag-reader}
```

```
$ clj-kondo --lint src:resources
resources/data_readers.clj:1:9: error: #'private-readers.core/my-tag-reader is private
src/private_readers/core.clj:2:8: warning: Unused private var private-readers.core/my-tag-reader
linting took 13ms, errors: 1, warnings: 1
```

**config**

N/A.

**expected behavior**

- Reference to private vars in `data_readers.clj` is not reported as an error, because Clojure allows doing that and the reader works correctly with private functions.
- A private function referenced in `data_readers.clj` is not considered unused.
- Maybe a way to exclude specific resource files in the `.clj-kondo/config.edn`.

**Comment by borkdude** [comment]
I'll fix the private access warning from data readers, but unused private is more complicated since it requires a global project state. I recommend just using an ignore hint here for now:

```
#_{:clj-kondo/ignore [:unused-private-var] :reason "used in data-readers.clj"}
(defn- my-private-reader [_]
  1)
```

**Comment by borkdude** [comment]
The same you can do in data_readers.clj for now:

```
#_{:clj-kondo/ignore [:private-call]}
```

**Comment by borkdude** [comment]
I've looked into this and while I'll be able to get rid of the private call warning in data_readers.cljc, the unused private var warning in the code itself can't be dealt with since this linter relies on usage within the same file.

> Using a private function as reader is useful when a library doesn't want to publicly expose the reader function as part of the API.

There are other ways to deal with this. E.g. call your function:

```
(defn ^:no-doc ^:skip-wiki my-reader-internal
  "INTERNAL: do not use"
  [x] x)
```

Using the word `internal`, adding `^:skip-wiki` and ` ^:no-doc` metadata and saying in the docstring that the function is only for internal use should be enough to communicate to humans and tooling that the var is not part of the public API.
